addons/mail/static/src/web/messaging_menu/messaging_menu.js

```diff
@@ -113,47 +111,42 @@ export class MessagingMenu extends Component {
         }
-        threads.sort((a, b) => {
-            if (a.correspondent === this.store.odoobot) {
+        return threads.sort((a, b) => {
+            /**
+             * Ordering:
+             * - threads with needaction
+             * - unread channels
+             * - read channels
+             * - odoobot chat
+             *
+             * In each group, thread with most recent message comes first
+             */
+            if (a.correspondent === this.store.odoobot && b.correspondent !== this.store.odoobot) {
                 return 1;
             }
-            if (!a.newestPersistentMessage?.datetime) {
+            if (a.hasNeedactionMessages && !b.hasNeedactionMessages) {
                 return -1;
             }
-            if (!b.newestPersistentMessage?.datetime) {
+            if (b.hasNeedactionMessages && !a.hasNeedactionMessages) {
                 return 1;
             }
-            return b.newestPersistentMessage.datetime - a.newestPersistentMessage.datetime;
-        });
-        const previews = [];
-        for (const thread of threads) {
-            const { newestMessage, newestNeedactionMessage } = thread;
-            if (thread.is_pinned) {
-                const message = newestMessage;
-                previews.push({
-                    id: `preview-${thread.localId}`,
-                    count: thread.message_unread_counter,
-                    imgUrl: thread.imgUrl,
-                    hasMarkAsReadButton: thread.message_unread_counter > 0,
-                    message,
-                    thread,
-                    isNeedaction: false,
-                });
+            if (a.message_unread_counter > 0 && b.message_unread_counter === 0) {
+                return -1;
             }
-            if (newestNeedactionMessage) {
-                const message = newestNeedactionMessage;
-                previews.push({
-                    id: `preview-needaction-${thread.localId}`,
-                    count: thread.needactionMessages.length,
-                    imgUrl: message.module_icon,
-                    hasMarkAsReadButton: true,
-                    message,
-                    thread,
-                    isNeedaction: true,
-                });
+            if (b.message_unread_counter > 0 && a.message_unread_counter === 0) {
+                return 1;
             }
-        }
-        return previews;
+            if (!a.newestPersistentMessage?.datetime && b.newestPersistentMessage?.datetime) {
+                return 1;
+            }
+            if (!b.newestPersistentMessage?.datetime && a.newestPersistentMessage?.datetime) {
+                return -1;
+            }
+            if (a.newestPersistentMessage?.datetime && b.newestPersistentMessage?.datetime) {
+                return b.newestPersistentMessage.datetime - a.newestPersistentMessage.datetime;
+            }
+            return b.id > a.id ? 1 : -1;
```

Comment:
Just using id won't make the result deterministic because model can be different. We need string compare on local id.